daemon/daemon.go

```diff
@@ -1199,9 +1201,118 @@ func (daemon *Daemon) setHostConfig(container *Container, hostConfig *runconfig.
 	if err := daemon.RegisterLinks(container, hostConfig); err != nil {
 		return err
 	}
-
+	var (
+		volumes []*vv
+		binds   = make(map[string]struct{})
+	)
+	for _, v := range hostConfig.VolumesFrom {
+		containerID, mode, err := parseVolumesFrom(v)
+		if err != nil {
+			return err
+		}
+		c, err := daemon.Get(containerID)
+		if err != nil {
+			return err
+		}
+		vf, err := container.addVolumesFrom(c, mode)
+		if err != nil {
+			return err
+		}
+		volumes = append(volumes, vf...)
+	}
+	hasVolumeFrom := func(dest string) bool {
+		for _, vf := range volumes {
+			if vf.vc.Destination == dest {
+				return true
+			}
+		}
+		return false
+	}
+	var (
+		uniqueVolumes       []volume.Volume
+		uniqueVolumeConfigs = make(map[string]*VolumeConfig)
+	)
+	for _, v := range container.volumes {
+		config := container.VolumeConfig[v.Name()]
+		if !hasVolumeFrom(config.Destination) {
+			uniqueVolumes = append(uniqueVolumes, v)
+			uniqueVolumeConfigs[v.Name()] = config
+		}
+	}
+	container.volumes = uniqueVolumes
+	container.VolumeConfig = uniqueVolumeConfigs
+	for _, vf := range volumes {
+		if _, exists := container.VolumeConfig[vf.v.Name()]; exists {
+			continue
+		}
+		container.volumes = append(container.volumes, vf.v)
+		container.VolumeConfig[vf.v.Name()] = vf.vc
+	}
+	uniqueVolumes = []volume.Volume{}
+	uniqueVolumeConfigs = make(map[string]*VolumeConfig)
+	// create bind mounts for container.
+	for _, b := range hostConfig.Binds {
+		// #10618
+		bind, err := parseBindMount(b)
+		if err != nil {
+			return err
+		}
+		if _, exists := binds[bind.Destination]; exists {
+			return fmt.Errorf("Duplicate bind mount %s", bind.Destination)
+		}
+		binds[bind.Destination] = struct{}{}
+		if _, err := os.Stat(bind.Source); err != nil {
+			if !os.IsNotExist(err) {
+				return err
+			}
+			if err := os.MkdirAll(bind.Source, 0755); err != nil {
+				return err
+			}
+		}
+		container.BindMounts = append(container.BindMounts, bind)
+	}
+	for _, v := range container.volumes {
+		config := container.VolumeConfig[v.Name()]
+		if !container.containsBind(config.Destination) {
+			uniqueVolumes = append(uniqueVolumes, v)
+			uniqueVolumeConfigs[v.Name()] = config
+		}
+	}
+	container.volumes = uniqueVolumes
+	container.VolumeConfig = uniqueVolumeConfigs
 	container.hostConfig = hostConfig
 	container.toDisk()
-
 	return nil
 }
+
+type vv struct {
+	v  volume.Volume
+	vc *VolumeConfig
+}
+
+func (container *Container) containsBind(dest string) bool {
+	for _, b := range container.BindMounts {
+		if b.Destination == dest {
+			return true
+		}
+	}
+	return false
+}
+
+func (container *Container) addVolumesFrom(c *Container, mode string) ([]*vv, error) {
+	var volumes []*vv
+	for name, config := range c.VolumeConfig {
+		lv, err := container.daemon.volumeDriver.Create(name)
```

Comment:
Wouldn't we want to look up the driver based on the container config of the volume?